### PR TITLE
adding a gulp task and getoptions

### PR DESCRIPTION
**Pull request by scottnath**
@sergesemashko @beynya 

putting together a common way to write include-able gulp tasks

### CODE REVIEW

**Comment by scottnath** [comment]
@sergesemashko @beynya 

Another pull request that is amazing...

File: lib/gulp-tasks/browsersync.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ */
+function getDefaultOptions () {
+
+  /* default options for pattern-importer */
+  var options = {
+
+    environment: {
+      config: {
+        server: {
+          baseDir: 'app/'
```
Comment (sergesemashko) [inline lib/gulp-tasks/browsersync.js]: In long run, I would suggest to do not have default value for this and throw an exception if it's not provided by callee.

Comment (scottnath) [inline lib/gulp-tasks/browsersync.js]: what if we made the default `./`? I'd like to have anything we make work if the end-user doesn't inject any options at all.

File: lib/gulp-tasks/file-globbing.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+/**
```
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: I took me few minutes and I had to read all the file to figure out what "globbing" means. Maybe there is a better name? I don't know, like "gulp-inject-task"

File: lib/get-options.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * @param	{object}	projectOptions	the project's options
+ */
+exports.getOptions = function (options, projectOptions) {
+
+  /* merge project and default options */
+  merge(options, projectOptions, function (a, b) {
+    return Array.isArray(a) ? b : undefined;
+  });
+	
+	if(!Array.isArray(options.dependencies)){
```
Comment (sergesemashko) [inline lib/get-options.js]: fix the indents

File: lib/gulp-tasks/file-globbing.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+ * @param {Object} gulp including file should in inject the gulp module
+ * @param {Object} projectOptions  object of options
+ * 
+ * @requires NPM:gulp-inject
+ */
+exports.gulpFileGlobbing = function (gulp, projectOptions) {
+
+  var options = getOptions(getDefaultOptions(),projectOptions);
+
+	gulp.task('file-globbing', options.dependencies, function () {
```
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: fix indents

File: lib/gulp-tasks/file-globbing.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+		var sources = gulp.src(options.files);
```
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: indents

File: lib/gulp-tasks/file-globbing.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ * @example
+ * // include in gulpfile.js
+ * require('pattern-library-utilities').gulpScssGlobbing(require('gulp'),scssGlobbingOptions);
+ * 
+ * @param {Object} gulp including file should in inject the gulp module
+ * @param {Object} projectOptions  object of options
+ * 
+ * @requires NPM:gulp-inject
+ */
+exports.gulpScssGlobbing = function (gulp, projectOptions) {
```
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: `gulpScssGlobbing ` function is almost the same as `gulpFileGlobbing` except `getDefaultScssOptions` call. can `gulpScssGlobbing` will be a wrapper of `gulpFileGlobbing`?

File: lib/gulp-tasks/browsersync.js (hunk 1 above)
Comment (sergesemashko) [inline lib/gulp-tasks/browsersync.js]: `./` is fine, or `app` if it is documented somewhere. `app/` looks just too related to our system on my opinion

File: lib/gulp-tasks/file-globbing.js (hunk 2 above)
Comment (scottnath) [inline lib/gulp-tasks/file-globbing.js]: @sergesemashko I see the point, as globbing is a relatively unknown term, but it is the *technical* term. And if a person was to search for a script that does this type of functionality, they would have much better luck using `globbing` or `glob` than they would if they used `inject`.

In node, the primary module which does this, which all others seem to be build upon, is called [node-glob](https://github.com/isaacs/node-glob). I'd like to keep this as `file-globbing`

@beynya debate?

File: lib/gulp-tasks/file-globbing.js (hunk 6 above)
Comment (scottnath) [inline lib/gulp-tasks/file-globbing.js]: @sergesemashko great idea!

File: lib/gulp-tasks/file-globbing.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ * Function to get default options for an implementation of gulp-inject
+ * @function
+ * @name getDefaultOptions
+ * 
+ * @returns {Object}  options  an object of options
+ */
+function getDefaultOptions () {
+
+  var options = {
+    config: {
```
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: could you document options?
It's not clear why src is a file and dest is not, what is dependencies responsible for, what is config.relative used for and etc

File: lib/gulp-tasks/browsersync.js (hunk 1 above)
Comment (scottnath) [inline lib/gulp-tasks/browsersync.js]: totally agree. I'll switch to `./`

File: lib/gulp-tasks/file-globbing.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+ * Function to get default options for globbing scss files with gulp-inject
+ * @function
+ * @name getDefaultScssOptions
+ * 
+ * @returns {Object}  options  an object of default options
+ */
+function getDefaultScssOptions () {
+
+  var options = {
+    config: {
```
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: I think `config` should be renamed to `injectRules`

**Comment by sergesemashko** [comment]
I suggest to have a eslint for all node modules we are writing, to check things like indents, unused variable and etc.

**Comment by sergesemashko** [comment]
@scottnath, quick question: Could you clarify the purpose for writing gulp tasks in separate modules? Why don't we have all gulp app related tasks in one folder like it's structured in atlas/gulp?

File: lib/gulp-tasks/file-globbing.js (hunk 2 above)
Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: I would say this file does more things than just globbing. It takes files and injects them into html file. Globbing means just to select files by specific pattern using stars like `app/*.js`,`**/*.js`.

**Comment by scottnath** [comment]
@sergesemashko I'm trying to make as much of the functionality that we build to be easily available in other apps. This stands for gulp tasks as well as any node modules we as well as...well, anything. The idea being to reduce the inevitable copy-n-paste that we will be forced to do if we want to use functionality like this outside of Atlas.
I'm looking to create the same mind-set we have with the pattern-library, that we can create code that can be used in multiple projects via include-and-use rather than copy-n-paste.

File: lib/gulp-tasks/file-globbing.js (hunk 2 above)
Comment (scottnath) [inline lib/gulp-tasks/file-globbing.js]: AHA. Very good counter-argument. Hmm. `fileGlobInject`?

Comment (sergesemashko) [inline lib/gulp-tasks/file-globbing.js]: I suggest to have pattern module names like `gulp-<task-name>-task`, so we know it contains gulp tasks.
`gulp-glob-inject-task`?

File: lib/gulp-tasks/file-globbing.js (hunk 8 above)
Comment (scottnath) [inline lib/gulp-tasks/file-globbing.js]: I'd like to leave this as it's the common name I've been using for module configs on other gulp-tasks

**Comment by scottnath** [comment]
@sergesemashko changes added

**Comment by sergesemashko** [comment]
looks good

**Comment by scottnath** [comment]
Thanks! will merge...

**Pull request merged**